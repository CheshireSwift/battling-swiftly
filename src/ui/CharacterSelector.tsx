import * as React from 'react'
import * as _ from 'lodash'

import Character from '../data/Character'

export type CharacterSelectorProps = {
  characters: Character[]
  availableKeys: string[]
  selected: string
  onSelect: (key: string) => void
}

export const CharacterSelector = ({
  characters,
  availableKeys,
  selected,
  onSelect,
}: CharacterSelectorProps) => (
  <div
    style={{
      position: 'fixed',
      background: 'black',
      color: 'lime',
      top: 0,
      right: 0,
    }}
  >
    {availableKeys.map(key => (
      <div
        id={'character-selector-option-' + key}
        key={key}
        style={{
          padding: '0.5rem',
          border: key === selected ? '1px solid lime' : '1px solid black',
        }}
        onClick={() => onSelect(key)}
      >
        {_.get(_.find(characters, { key }), 'name')}
      </div>
    ))}
  </div>
)

export default CharacterSelector
