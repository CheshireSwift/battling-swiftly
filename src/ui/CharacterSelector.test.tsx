import * as React from 'react'
import { render, mount } from 'enzyme'
import CharacterSelector, { CharacterSelectorProps } from './CharacterSelector'

describe('the character selector', () => {
  const characterSelector = (props?: Partial<CharacterSelectorProps>) => (
    <CharacterSelector
      characters={[
        { key: 'hero', name: 'Waffle', position: { x: 0, y: 0 } },
        { key: 'villain', name: 'Pickle', position: { x: 1, y: 1 } },
        { key: 'misc', name: 'Blep', position: { x: 2, y: 2 } },
      ]}
      availableKeys={['hero', 'villain']}
      selected={'villain'}
      onSelect={jest.fn()}
      {...props}
    />
  )

  it('matches the snapshot', () => {
    expect(render(characterSelector())).toMatchSnapshot()
  })

  it('displays the character for each key that is available to select', () => {
    const selectorText = render(characterSelector()).text()
    expect(selectorText).toContain('Waffle')
    expect(selectorText).toContain('Pickle')
    expect(selectorText).not.toContain('Blep')
  })

  it('highlights the selected character', () => {
    const selector = render(characterSelector())
    const heroOptionStyle = selector
      .find('#character-selector-option-hero')
      .attr('style')
    const villainOptionStyle = selector
      .find('#character-selector-option-villain')
      .attr('style')

    expect(villainOptionStyle).not.toEqual(heroOptionStyle)
  })

  it('fires onSelect when a new character is selected', () => {
    const onSelect = jest.fn()
    const selector = mount(characterSelector({ onSelect }))
    selector.find('#character-selector-option-hero').simulate('click')
    expect(onSelect).toHaveBeenCalledWith('hero')
  })
})
