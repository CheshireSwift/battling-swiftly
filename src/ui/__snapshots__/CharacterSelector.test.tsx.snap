// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the character selector matches the snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "style": "position:fixed;background:black;color:lime;top:0;right:0",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "id": "character-selector-option-hero",
          "style": "padding:0.5rem;border:1px solid black",
        },
        "children": Array [
          Object {
            "data": "Waffle",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "id": "character-selector-option-villain",
            "style": "padding:0.5rem;border:1px solid lime",
          },
          "children": Array [
            Object {
              "data": "Pickle",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "id": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "id": undefined,
            "style": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "id": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "id": undefined,
          "style": undefined,
        },
      },
      Object {
        "attribs": Object {
          "id": "character-selector-option-villain",
          "style": "padding:0.5rem;border:1px solid lime",
        },
        "children": Array [
          Object {
            "data": "Pickle",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "id": "character-selector-option-hero",
            "style": "padding:0.5rem;border:1px solid black",
          },
          "children": Array [
            Object {
              "data": "Waffle",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "id": undefined,
            "style": undefined,
          },
          "x-attribsPrefix": Object {
            "id": undefined,
            "style": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "id": undefined,
          "style": undefined,
        },
        "x-attribsPrefix": Object {
          "id": undefined,
          "style": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "style": undefined,
    },
    "x-attribsPrefix": Object {
      "style": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
